pandasai/llm/gpt4all.py

```diff
@@ -0,0 +1,105 @@
+LLMs maintained by Nomic AI. With GPT4All we can easily download models like 
+Falcon, Llama etc and run them seamlessly using CPU with lower RAMs. 
+
+Example:
+    Use below example to call GPT4All supporrted models
+
+    >>> from pandasai.llm.gpt4all import GPT4AllLLM
+"""
+import requests
+from typing import Optional
+from gpt4all import GPT4All
+
+from pandasai.prompts.base import Prompt
+from .base import LLM
+
+
+class GPT4AllLLM(LLM):
+    """GPT4All LLMs API
+    Base LLM class is extended to support GPT4AllLLM. When this class will be
+    initialized all the additional parameters like temp, top_p, top_k etc should
+    be inside **kwargs while instantiating the class. That is to be done only if
+    the user wants to override the existing configurations. Below example shows how
+    we can use override certain configurations to change model's behaviour.
+
+    Example:
+        >>> import pandas as pd
+        >>> from pandasai import PandasAI
+        >>> from pandasai.llm.gpt4all import GPT4AllLLM
+
+        >>> model_name = 'ggml-replit-code-v1-3b.bin'
+        >>> model_path = 'path where the model is loaded or to be downloaded'
+        >>> additional_params = {'temp': 1, 'max_tokens': 50}
+        >>> model = GPT4AllLLM(model_folder_path, model_name, **additional_params)
+
+        >>> df_ai = PandasAI(model)
+        >>> response = df_ai(df, prompt='What is the sum of the GDP in this table?')
+
+    Note: Please note that right now Pandas AI only supports models for gpt4all. However
+    it might not work as chatGPT when it comes to performance, hence for now users using
+    this module have to tune the existing prompts to get better results.
+    """
+
+    temp: Optional[float] = 0.7
+    top_p: Optional[float] = 0.1
+    top_k: Optional[int] = 40
+    n_batch: Optional[int] = 8
+    n_threads: Optional[int] = 4
+    n_predict: Optional[int] = 256
+    max_tokens: Optional[int] = 200
+    repeat_last_n: Optional[int] = 64
+    repeat_penalty: Optional[float] = 1.18
+
+    _model_repo_url = "https://gpt4all.io/models/models.json"
+    _supported_models = [
+        metadata["filename"] for metadata in requests.get(_model_repo_url).json()
+    ]
+
+    def __init__(
+        self,
+        model_folder_path: str,
+        model_name: str,
+        allow_download: Optional[bool] = None,
+        n_threads: Optional[bool] = None,
+        **kwargs,
+    ) -> None:
+        self.model_name = (
+            f"{model_name}.bin" if not model_name.endswith(".bin") else model_name
+        )
+        assert_msg = (
+            "Model not Found. Please check the list of supported models:",
+            self._supported_models,
+        )
+        assert self.model_name in self._supported_models, assert_msg
+
+        self.gpt4all_model = GPT4All(
+            model_name=self.model_name,
+            model_path=model_folder_path,
+            allow_download=allow_download,
+            n_threads=n_threads,
+        )
```

Comment:
Since this module requires `GPT4All`, you should either place it among the mandatory requirements of pandasai (not recommended) or trying importing it and catching the `ImportError` raising an Exception instructing the user to first run `pip install gpt4all`

Reply:
That's a good suggestion, will do that